### PR TITLE
Valid generated config acceptance test

### LINKED ISSUE #845: Test that generate produces a valid config
In our acceptance tests, we should test that generate produces a config that is accepted by start and seed.

### PR DESCRIPTION
**Pull request by oxarbitrage**
Closes https://github.com/ZcashFoundation/zebra/issues/845

### CODE REVIEW

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Suggestion: Use `PathBuf.push`, which will automatically add the correct path separators.

https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.push

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: Use the temporary directory for each test, not the shared build directory.

Using the shared build directory can lead to file read/write conflicts.

Here's how you can access the temporary directory:
```suggestion
    let tempdir_path = _guard.path();
```

`_guard` is actually a `TempDir` instance:
https://docs.rs/tempdir/0.3.7/tempdir/struct.TempDir.html#method.path

Suggestion: rename `_guard` to `tempdir`.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: If you use `PathBuf`, you can call `exists` directly:
https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.exists

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: See my suggestions above about `TempDir` and `PathBuf`.

Also, rather than duplicating this code, you might be able to put it in a separate function.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: Use `zebrad -c /path/to/config.toml` to make sure zebrad is actually reading the generated config.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for writing these tests!

We should use a temporary directory for the generated configs, and pass an explicit config path to zebrad.

I also suggest we use `PathBuf` to build paths, because it's easier.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I actually tried something like this but i had a problem. `_guard` is not just an instance of `TempDir` but a `impl std::ops::Drop` type that i don't know how to cast to a tempdir or path.  Suggested code ` let tempdir_path = _guard.path();` will fail to build.

One option that i tried is to return a `TempDir` third argument of `test_cmd`, that will work but ugly.

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Thanks, a lot better. Applied at https://github.com/ZcashFoundation/zebra/pull/859/commits/7a9bd4ae38c4bb8766e32bf656e8cf4c65a8e70c

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Added this at https://github.com/ZcashFoundation/zebra/pull/859/commits/18c4c5579adbc7b0346f1433176700fba4a97e03

File: zebrad/tests/acceptance.rs
Comment (yaahc) [reply zebrad/tests/acceptance.rs]: I think you're on the right path, we're going to have to tweek the return type to a concrete type rather than an existential so we can call fns on it. If you'd like you can try cleaning up the API by wrapping the tempdir's guard with new_type that provides a nicer interface

Something like:

```rust
struct ZebraTestDir {
    tempdir: TempDir,
}

impl ZebraTestDir {
    fn with_config(&self, cmd: &mut Command, config: Config) -> Result<()> {
        // write config to the directory and add the correct args to the `Command` to get it to use the config.
    }
}
```

Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: I added this at https://github.com/ZcashFoundation/zebra/pull/859/commits/3b8925405f84b153540ef0e36f3318bd567de703 however i am not totally convinced but pushed. Maybe with a re review of you guys we can make something decent. Thanks.

Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Please hold your review until tomorrow. Need adjustments.

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs]: I like this approach, but rather than having the arg that we set to `None` most of the time I think it might be good to just expose two versions of `test_cmd` and `get_child`, one for new command tests and one for new commands on in progress tests. I'm thinking `get_child` and `get_child_in`.

Honestly, you don't even have to change it here, its probably fine if you only update the `get_child` fns to have the two interfaces.

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs]: :+1:

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs]: `Result` has a helper for converting it into an option. I don't think we want to use it here, so if you already know this then feel free to ignore. In this case its probably better if we panic outright if the tempdir fails to create.

```rust
            tempdir: TempDir::new(prefix).ok(),
```

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: rather than unwrapping I'd use a pattern for the match
```suggestion
    assert_eq!(path_config.parent(), Some(tempdir.path()));
```

**Review by yaahc — CHANGES_REQUESTED** [review]
Looks great! I have one minor tweek I'd like to make that I think will clean up a lot of the usage but I think this is good to merge other than that.

**Comment by oxarbitrage** [comment]
I changed the approach completely, sorry about that, the PR is going to need a new full review.

Instead of creating the temp dir in `test_cmd()` i decided to separate the creation of this into a separated function `tempdir()` that is executed at the top of each test and then used in the whole test without losing the scope. Among other alternatives this was the one that convinced me more.

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs]: Instead of setting the current dir to the directory storing the binary we're testing we should pass in the path for the tempdir or even just a reference to the TempDir itself.

File: zebra-test/src/command.rs
Comment (yaahc) [inline zebra-test/src/command.rs:21]: Given that this pathbuf is usually ignored we should probably just return a `TempDir` instead of the `impl Drop` and let the tests get the path when they need it from the `TempDir` itself.

**Review by yaahc — COMMENTED** [review]
> Instead of creating the temp dir in `test_cmd()` i decided to separate the creation of this into a separated function `tempdir()` that is executed at the top of each test and then used in the whole test without losing the scope. Among other alternatives this was the one that convinced me more.

:+1:

File: zebra-test/src/command.rs
Comment (oxarbitrage) [reply zebra-test/src/command.rs]: https://github.com/ZcashFoundation/zebra/pull/859/commits/a27e5a243fb98084cf6ee8ce7852113661e5a9c5

**Review by yaahc — DISMISSED** [review]

File: zebra-test/src/command.rs
Comment (oxarbitrage) [reply zebra-test/src/command.rs:21]: By the changes done at https://github.com/ZcashFoundation/zebra/pull/859/commits/a27e5a243fb98084cf6ee8ce7852113661e5a9c5 the pathbuf is now actually used. Not sure if we still want to make changes here.

**Comment by yaahc** [comment]
looks like a recent merge introduced a conflict but other than that I'm happy with this, thanks as always @oxarbitrage.

**Review by yaahc — DISMISSED** [review]

**Comment by oxarbitrage** [comment]
Conflict was with https://github.com/ZcashFoundation/zebra/pull/887 specifically the PR added `memory_cache_bytes = 256000000` to the default generated config. I fixed this using the github so waiting for the CI to see if it worked.

**Comment by hdevalence** [comment]
ah that's probably my fault — I think the state Config should have whatever attribute makes serde load unspecified fields from defaults

File: zebra-test/src/command.rs
Comment (teor2345) [inline zebra-test/src/command.rs:21]: This is a lot cleaner, and easier to understand. Thanks!

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: On this line, tempdir isn't a directory any more.

So let's clone it, push the config file name, and use a different variable name for the generated config file.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: Let's use a different variable name for the generated config file.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: Let's use a different variable name for the generated config file.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks for these changes, they are a lot easier to understand.

Let's use a different name for the generated config file, because it's not a directory.

There is also a minor behavior change in this PR: since we only use one tempdir per test function, we re-use state from previous zebrad commands in the same test function. I don't think that's a problem: at worst, we will get repeated failures if we corrupt the state. (And if that happens, we should fix the corruption issue.)

File: zebrad/tests/acceptance.rs
Comment (oxarbitrage) [reply zebrad/tests/acceptance.rs]: Thanks, a lot better and easier to read. Done at https://github.com/ZcashFoundation/zebra/pull/859/commits/f9cc9bde4d7cefa1fc326c97f315ab766f5a607c

**Review by teor2345 — APPROVED** [review]
Thanks!

I think we're good to mere here, @yaahc did you want to do a final review?

**Review by yaahc — APPROVED** [review]
looks good to me! I'm gonna go ahead and merge this. :partying_face:

**Pull request merged**
